molgenis-data/src/main/java/org/molgenis/data/decorator/PermissionCheckingDecorator.java

```diff
@@ -0,0 +1,188 @@
+package org.molgenis.data.decorator;
+
+import org.molgenis.data.*;
+import org.molgenis.data.support.QueryImpl;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.StreamSupport.stream;
+
+/**
+ * Decorator that checks if actions on this repository are allowed using a PermissionChecker.
+ *
+ * @param <E> the class of the entities in the repository
+ */
+public class PermissionCheckingDecorator<E extends Entity> extends AbstractRepositoryDecorator<E>
```

Comment:
Do we have existing code that can be replaced with this new decorator? I remember seeing this code before, which if correct would mean introducing code duplication.